### ISSUE TITLE
full_control eating up memory

### ISSUE THREAD

**Issue by MakinoharaShouko** [osrf/rmf_core#129]
When I sent loop request in my own rmf demo scenario, I encountered a problem of the computer memory used up by process full_control.

Here is my [repo](https://github.com/MakinoharaShouko/rmf_demos). I am using the master branch. I added a hotel map and a launch file called hotel.launch.xml under demos to start the 3D simulation and the rviz panels to send the requests.

Here are steps to reproduce the problem:
1. Clone rmf_core, traffic_editor, rmf_scheduler_visualizer and my rmf_demos (all on master branches) into one workspace
2. Do `colcon build` and source `setup.bash` in `install`
3. Run `ros2 launch demos hotel.launch.xml`
4. In the rviz rmf panel, send loop request starting from room2, ending at room7 to fleet magni, repeating once

I also tried different requests with different versions of maps. The memory usage increases as the loop length increases. The number of doors also had a large impact on memory consumption. Additionally, I also noticed that the memory used by full_control did not go down even after the loop finished.

**Comment by mxgrey** [comment]
Yep, there is currently some kind of memory leak issue in the `full_control` fleet adapter that we are actively trying to track down. We haven't managed to recreate the leak in any unit tests yet, and it's not feasible to run valgrind alongside a simulation (because valgrind slows down the fleet adapter so badly that the system doesn't work at all), so this has been a particularly frustrating bug to track down.

We're hoping to find the source of the leak and patch it in the coming weeks, but until then there will be an unfortunate ceiling on how long a fleet adapter can continuously run.

**Comment by MakinoharaShouko** [comment]
> Yep, there is currently some kind of memory leak issue in the `full_control` fleet adapter that we are actively trying to track down. We haven't managed to recreate the leak in any unit tests yet, and it's not feasible to run valgrind alongside a simulation (because valgrind slows down the fleet adapter so badly that the system doesn't work at all), so this has been a particularly frustrating bug to track down.
> 
> We're hoping to find the source of the leak and patch it in the coming weeks, but until then there will be an unfortunate ceiling on how long a fleet adapter can continuously run.

I used valgrind to check any memory leak and the leak seemed pretty small. Sorry I forgot to mention it earlier
```
==10144== LEAK SUMMARY:
==10144==    definitely lost: 36 bytes in 2 blocks
==10144==    indirectly lost: 12 bytes in 1 blocks
==10144==      possibly lost: 0 bytes in 0 blocks
==10144==    still reachable: 71,281 bytes in 106 blocks
==10144==         suppressed: 0 bytes in 0 blocks
```

**Comment by mxgrey** [comment]
It's very possible that what I mean by "leak" won't even be caught by valgrind. There may be objects that are accidentally being kept alive even after their usefulness has expired, but those objects will still be freed when the entire executable gets torn down.

**Comment by MakinoharaShouko** [comment]
I pulled the latest rmf_core to see if my problem of memory used up is fixed by the new update. However, I have been getting the following errors when doing `colcon build`:
```
--- stderr: rmf_fleet_adapter                                
In file included from /home/makinohara/rmf_demos_gazebo_ws/src/rmf_core/rmf_fleet_adapter/src/rmf_fleet_adapter/services/Negotiate.hpp:135,
                 from /home/makinohara/rmf_demos_gazebo_ws/src/rmf_core/rmf_fleet_adapter/src/rmf_fleet_adapter/services/Negotiate.cpp:18:
/home/makinohara/rmf_demos_gazebo_ws/src/rmf_core/rmf_fleet_adapter/src/rmf_fleet_adapter/services/detail/impl_Negotiate.hpp: In lambda function:
/home/makinohara/rmf_demos_gazebo_ws/src/rmf_core/rmf_fleet_adapter/src/rmf_fleet_adapter/services/detail/impl_Negotiate.hpp:45:44: error: ‘using element_type = const class rmf_traffic::schedule::Negotiation::Table::Viewer’ {aka ‘const class rmf_traffic::schedule::Negotiation::Table::Viewer’} has no member named ‘defunct’; did you mean ‘sequence’?
     return *service_interrupted || viewer->defunct();
                                            ^~~~~~~
                                            sequence
/home/makinohara/rmf_demos_gazebo_ws/src/rmf_core/rmf_fleet_adapter/src/rmf_fleet_adapter/services/detail/impl_Negotiate.hpp: In member function ‘void rmf_fleet_adapter::services::Negotiate::operator()(const Subscriber&)’:
/home/makinohara/rmf_demos_gazebo_ws/src/rmf_core/rmf_fleet_adapter/src/rmf_fleet_adapter/services/detail/impl_Negotiate.hpp:55:14: error: ‘using Options = class rmf_traffic::agv::Planner::Options’ {aka ‘class rmf_traffic::agv::Planner::Options’} has no member named ‘interrupter’; did you mean ‘interrupt_flag’?
             .interrupter(interrupter));
              ^~~~~~~~~~~
              interrupt_flag
```
I tried to look at the code where the error happens, but I found the members are all defined. So I am now really confused why these errors are there.

**Comment by mxgrey** [comment]
The only thing I can think of is maybe you did

```
$ source /home/makinohara/rmf_demos_gazebo_ws/install/setup.bash
```

before trying to build, and so the compiler is trying to use the old header definitions instead of the new ones. If that's the case, then you should open up a new terminal and source only the Eloquent distribution:

```
source /opt/ros/eloquent/setup.bash
```

If the compiler error persists after that, then I don't know what the problem is, and I'd have to learn more about your setup.

**Comment by MakinoharaShouko** [comment]
Actually I never wrote that line to my .bashrc and did not source before building either.

I also commented out all the source lines of ROS2 except `source /opt/ros/eloquent/setup.bash` and removed the folders from the previous building in order to eliminate all possible influences, but the problem still exists.

Here is my .bashrc file. Not sure if it would be useful
```
# ~/.bashrc: executed by bash(1) for non-login shells.
# see /usr/share/doc/bash/examples/startup-files (in the package bash-doc)
# for examples

# If not running interactively, don't do anything
case $- in
    *i*) ;;
      *) return;;
esac

# don't put duplicate lines or lines starting with space in the history.
# See bash(1) for more options
HISTCONTROL=ignoreboth

# append to the history file, don't overwrite it
shopt -s histappend

# for setting history length see HISTSIZE and HISTFILESIZE in bash(1)
HISTSIZE=1000
HISTFILESIZE=2000

# check the window size after each command and, if necessary,
# update the values of LINES and COLUMNS.
shopt -s checkwinsize

# If set, the pattern "**" used in a pathname expansion context will
# match all files and zero or more directories and subdirectories.
#shopt -s globstar

# make less more friendly for non-text input files, see lesspipe(1)
[ -x /usr/bin/lesspipe ] && eval "$(SHELL=/bin/sh lesspipe)"

# set variable identifying the chroot you work in (used in the prompt below)
if [ -z "${debian_chroot:-}" ] && [ -r /etc/debian_chroot ]; then
    debian_chroot=$(cat /etc/debian_chroot)
fi

# set a fancy prompt (non-color, unless we know we "want" color)
case "$TERM" in
    xterm-color|*-256color) color_prompt=yes;;
esac

# uncomment for a colored prompt, if the terminal has the capability; turned
# off by default to not distract the user: the focus in a terminal window
# should be on the output of commands, not on the prompt
#force_color_prompt=yes

if [ -n "$force_color_prompt" ]; then
    if [ -x /usr/bin/tput ] && tput setaf 1 >&/dev/null; then
# We have color support; assume it's compliant with Ecma-48
# (ISO/IEC-6429). (Lack of such support is extremely rare, and such
# a case would tend to support setf rather than setaf.)
color_prompt=yes
    else
color_prompt=
    fi
fi

if [ "$color_prompt" = yes ]; then
    PS1='${debian_chroot:+($debian_chroot)}\[\033[01;32m\]\u@\h\[\033[00m\]:\[\033[01;34m\]\w\[\033[00m\]\$ '
else
    PS1='${debian_chroot:+($debian_chroot)}\u@\h:\w\$ '
fi
unset color_prompt force_color_prompt

# If this is an xterm set the title to user@host:dir
case "$TERM" in
xterm*|rxvt*)
    PS1="\[\e]0;${debian_chroot:+($debian_chroot)}\u@\h: \w\a\]$PS1"
    ;;
*)
    ;;
esac

# enable color support of ls and also add handy aliases
if [ -x /usr/bin/dircolors ]; then
    test -r ~/.dircolors && eval "$(dircolors -b ~/.dircolors)" || eval "$(dircolors -b)"
    alias ls='ls --color=auto'
    #alias dir='dir --color=auto'
    #alias vdir='vdir --color=auto'

    alias grep='grep --color=auto'
    alias fgrep='fgrep --color=auto'
    alias egrep='egrep --color=auto'
fi

# colored GCC warnings and errors
#export GCC_COLORS='error=01;31:warning=01;35:note=01;36:caret=01;32:locus=01:quote=01'

# some more ls aliases
alias ll='ls -alF'
alias la='ls -A'
alias l='ls -CF'

# Add an "alert" alias for long running commands.  Use like so:
#   sleep 10; alert
alias alert='notify-send --urgency=low -i "$([ $? = 0 ] && echo terminal || echo error)" "$(history|tail -n1|sed -e '\''s/^\s*[0-9]\+\s*//;s/[;&|]\s*alert$//'\'')"'

# Alias definitions.
# You may want to put all your additions into a separate file like
# ~/.bash_aliases, instead of adding them here directly.
# See /usr/share/doc/bash-doc/examples in the bash-doc package.

if [ -f ~/.bash_aliases ]; then
    . ~/.bash_aliases
fi

# enable programmable completion features (you don't need to enable
# this, if it's already enabled in /etc/bash.bashrc and /etc/profile
# sources /etc/bash.bashrc).
if ! shopt -oq posix; then
  if [ -f /usr/share/bash-completion/bash_completion ]; then
    . /usr/share/bash-completion/bash_completion
  elif [ -f /etc/bash_completion ]; then
    . /etc/bash_completion
  fi
fi

source /usr/share/gazebo/setup.sh

export IGN_CONFIG_PATH=$IGN_CONFIG_PATH:/usr/share/ignition

source /opt/ros/eloquent/setup.bash
#source ~/bridge_ws/install/setup.bash

PYTHONWARNINGS=ignore::yaml.YAMLLoadWarning

alias g++="g++ -std=c++11"
#source ~/ws_dome/install/setup.bash

```

**Comment by mxgrey** [comment]
Then the only thing I can think of suggesting is to wipe out the build and install directories of your workspace:

```
cd /home/makinohara/rmf_demos_gazebo_ws
rm -rf build install
```

And then start over from a fresh terminal. If that doesn't work, then something very very suspicious is happening in your `src` folder.

**Comment by MakinoharaShouko** [comment]
I just tried the new rmf_come and the problem of memory being used up still exists, though I saw it went down to the normal level if the route finding was able to finish.

Additionally, in my previous comment, I said that the memory used is related to the loop length, but I just realized that it does not seem to be that way. The memory usage remained the same for long loops and short loops in my most recent trials.

Now I am really confused that the airport terminal is almost the same burden as my hotel map but it does not experience such a high usage of memory.

**Comment by mxgrey** [comment]
Yeah, some funny things are definitely happening in the planner. It's technically producing correct plans, but the amount of graph searching it's doing is way way higher than it should have to be. I think a combination of caching, improving the heuristic, and intelligent pruning should bring down the memory consumption and planning time by several orders of magnitude. That'll be my highest priority as soon as I've finished some urgent feature requests.

**Comment by mxgrey** [comment]
This issue should be largely resolved by #243. There is still room to improve the memory consumption of the planner, and we have plans to do so, but the severity of this problem should be largely mitigated now.
